src/core/utils/qfieldcloudutils.cpp

```diff
@@ -86,3 +89,70 @@ const QVariant QFieldCloudUtils::projectSetting( const QString &projectId, const
   const QString projectPrefix = QStringLiteral( "QFieldCloud/projects/%1" ).arg( projectId );
   return settings.value( QStringLiteral( "%1/%2" ).arg( projectPrefix, setting ), defaultValue );
 }
+
+const QMultiMap<QString, QString> QFieldCloudUtils::getPendingAttachments()
+{
+  QMultiMap<QString, QString> files;
+
+  QLockFile attachmentsLock( QStringLiteral( "%1/attachments.lock" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+  if ( attachmentsLock.tryLock( 10000 ) )
+  {
+    QFile attachmentsFile( QStringLiteral( "%1/attachments.csv" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+    QFileInfo fi( attachmentsFile );
+    if ( !fi.exists() || fi.size() == 0 )
+      return std::move( files );
+
+    attachmentsFile.open( QFile::ReadWrite | QFile::Text );
+    QTextStream attachmentsStream( &attachmentsFile );
+    while ( !attachmentsStream.atEnd() )
+    {
+      const QString line = attachmentsStream.readLine().trimmed();
+      const QStringList values = StringUtils::csvToStringList( line );
+
+      if ( values.size() >= 2 )
+      {
+        files.insert( values.at( 0 ), values.at( 1 ) );
+      }
+    }
+  }
+  return std::move( files );
+}
+
+void QFieldCloudUtils::addPendingAttachment( const QString &projectId, const QString &fileName )
+{
+  QLockFile attachmentsLock( QStringLiteral( "%1/attachments.lock" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+  if ( attachmentsLock.tryLock( 10000 ) )
+  {
+    QStringList values = QStringList() << projectId << fileName;
+
+    QFile attachmentsFile( QStringLiteral( "%1/attachments.csv" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+    attachmentsFile.open( QFile::Append | QFile::Text );
+    QTextStream attachmentsStream( &attachmentsFile );
+    attachmentsStream << Qt::endl
+                      << StringUtils::stringListToCsv( values );
```

Comment:
```suggestion
    attachmentsStream << StringUtils::stringListToCsv( values )
                      << Qt::endl;
```
hm, this is going to create an empty line when we first insert a file here, no? I think a better practice is to always leave the last line empty, so just reverse the order of these chunks you send to the stream.

Reply:
And super slight optimization, when you parse the file, just add a check if consumed line == "" (empty string).